Util/Docker/Carla.Dockerfile-mod

```diff
@@ -0,0 +1,39 @@
+FROM carla-prerequisites:latest as build-env
+
+ARG CARLA_REPO=https://github.com/Bonifatius94/carla
+ARG CARLA_BRANCH=keep-scoomatic-resources-during-update
+
+USER carla
+WORKDIR /home/carla
+
+# TODO: don't clone, use the locally checked out repo via ADD command instead
+RUN if [ -z ${CARLA_BRANCH+x} ]; then git clone --depth 1 $CARLA_REPO; \
+    else git clone --depth 1 --branch $CARLA_BRANCH $CARLA_REPO; fi
+
+WORKDIR /home/carla/carla
+RUN ./Update.sh
+RUN make CarlaUE4Editor
+RUN make PythonAPI
+RUN make build.utils
+RUN make package
+# TODO: chain this in one RUN and remove the temporary build files afterwards
+#       -> make the docker layer smaller
+
+FROM nvidia/vulkan:1.1.121-cuda-10.1--ubuntu18.04 as runtime
+
+RUN useradd -m carla
+WORKDIR /home/carla
+COPY --from=build-env --chown=carla:carla /home/carla/carla/Dist/CARLA_Shipping_*/LinuxNoEditor .
```

Comment:
Can you help me to understand what is happening here?

build-env is your first build stage and runtime is your second stage. I would have expected, that `carla` is already a user of `build-env`. Why do you have to add carla again? Then you copy the `LinuxNoEditor` from the first build stage to the second build stage? And you change the owner, without changing the owner? I am confused.

Reply:
Build stages are like a new image. Everything from previous stages gets discarded unless you tell it to copy stuff over. That's also the reason why the output image is only 16GB although the build context is 160GB.

Reply:
So all files are discarded, but the software environment from the previous stage persists?

Reply:
Yes, the Docker layers are cached in case you wanna rebuild with the same inputs to speed up builds in a local dev environment with potentially constrained computing resources.

In case it's a server-side CI/CD pipeline, the intermediate layers are usually discarded once the final layer got published to a Docker registry, that is to ensure all commands are still working. But there are other scenarios where lots of devs are running the pipeline for CI with other dev's changes. So usually the layers need to be cached during worktime for pipeline responses within 5-10 minutes and the cache is only rebuild once over night.

And there's also the notion of mono-repos where lots of components are stored within a giant repo (e.g. Google does this). In such a mono-repo, the CI/CD pipeline is a lot more intelligent in a way that it recognizes which components are affected by a change and only rebuilds those. But this doesn't apply to our scenario because we're stuck with a giant monolith ^^